### ISSUE TITLE
Avoid volatile field in Immutable generated code (Question)

### ISSUE THREAD

**Issue by chevaris** [immutables/immutables#1366]
Hi there,

First, really thanks for this amazing piece of code that makes our life much easier!!!

When @Value.default annotation is used, generated classes include a volatile field and inner class (InitShim).

I am not 100% sure this is a good idea for several reasons:
- a volatile field breaks the canonical definition of immutable class for Java memory model, because potentially allows field mutation after construction and avoids safe publication
- getters for fields using annotation require a null check for the volatile field, on the assumption that the object could accessed when the object is not fully constructed. Even if neglictable, there is minor performance glitch. In general, I assume that Immutables classes can be used in high concurrent servers and one of the befits is maximize performance glitches.

In my view nothing prevents generate code for the classes in which everything can be done in class constructor without the support of a volatile field. The main benefits are:
- Immutable class is a perfect match of the immutable pattern
- Volatile field is avoided and null check is avoided in getters

I am 100% sure that you have already discussed about this topic when design was done, and I would appreciate if you could share your arguments

Regards,

/Chevaris

**Comment by elucash** [comment]
TL;DR there's `Style.unsafeDefaultAndDerived = true` will revert to the old ways where there's no `initShim`

* During construction, default and derived have their "initializer" method bodies called to get a value, these bodies can refer to other attributes which in turn can be default or derived. This will work as long as there is no cycles in dependencies (exception at runtime, but not a stack overflow).
* InitShim is the price for this functionality and safety. Having final and volatile field in this case is memory safe and provides safe publishing. The only this you might imagine is that is not as efficient as all-final due to additional memory barrier.
* If you're sure that you will never need extra safety while calling attributes from default or derived initialized bodies and sure to never call not yet initialized ones, then you can skip InitShim altogether by using `Style.unsafeDefaultAndDerived = true` style flag

**Comment by chevaris** [comment]
Thanks a lot for the explanation and the tip for the Style flag
